preprocessing/science/refine_srf.py

```diff
@@ -0,0 +1,174 @@
+            print("no plane specified")
+            raise
+        line_el = fid.readline().split()
+        nx = int(line_el[2])
+        ny = int(line_el[3])
+        line_el = fid.readline().split()
+        if line_el[0] != "POINTS":
+            print("only one plane supported")
+            raise
+        # check that the plane data are consistent with the number of points
+        assert int(line_el[1]) == nx * ny
+        p1 = FaultPlane(nx, ny)
+        for j in range(ny):
+            for i in range(nx):
+                # first header line
+                line = fid.readline()
+                p1.lon[j, i], p1.lat[j, i], p1.depth[j, i], p1.strike[j, i], p1.dip[j, i], p1.PSarea, p1.t0[j, i], p1.dt = [float(v) for v in line.split()]
+                # second header line
+                line = fid.readline()
+                p1.rake[j, i], p1.slip1[j, i], ndt1, slip2, ndt2, slip3, ndt3 = [float(v) for v in line.split()]
+                ndt1 = int(ndt1)
+                if max(i, j) == 0:
+                    p1.ndt = ndt1
+                    p1.init_aSR()
+                lSTF = []
+                if ndt1 == 0:
+                    continue
+                if ndt1 > p1.ndt:
+                    print("this script assumes that ndt is not larger than the first ndt in the file", ndt1, p1.ndt)
+                    raise
+                while True:
+                    line = fid.readline()
+                    lSTF.extend(line.split())
+                    if len(lSTF) == ndt1:
+                        p1.aSR[j, i, 0:ndt1] = np.array([float(v) for v in lSTF])
+                        break
+    return p1
+
+
+def upsampleFault(p, spatial_order, spatial_zoom, temporal_zoom):
+    # time vector
+    ndt2 = (p.ndt - 1) * temporal_zoom + 1
+    ny2, nx2 = p.ny * spatial_zoom, p.nx * spatial_zoom
+    # resampled source
+    p2 = FaultPlane(nx2, ny2)
+    p2.ndt = ndt2
+    p2.init_aSR()
+
+    p2.dt = p.dt / temporal_zoom
+    p2.compute_time()
+
+    # upsample spatially all these quantitites
+    allarr = np.array([p.x, p.y, p.depth, p.t0, p.slip1, p.strike, p.dip, p.rake])
+    nd = allarr.shape[0]
+    allarr0 = np.zeros((nd, p2.ny, p2.nx))
+
+    for k in range(nd):
+        allarr0[k, :, :] = scipy.ndimage.zoom(allarr[k, :, :], spatial_zoom, order=spatial_order)
+    p2.x, p2.y, p2.depth, p2.t0, p2.slip1, p2.strike, p2.dip, p2.rake = allarr0
+    p2.compute_latlon_from_xy(args.proj)
+    p2.PSarea = p.PSarea / spatial_zoom ** 2
+
+    aSRa = np.zeros((p2.ny, p2.nx, p.ndt))
+    for k in range(p.ndt):
+        aSRa[:, :, k] = scipy.ndimage.zoom(p.aSR[:, :, k], spatial_zoom, order=spatial_order)
+
+    # interpolate temporally the AST
+    for j in range(p2.ny):
+        for i in range(p2.nx):
+            f = interpolate.interp1d(p.myt, aSRa[j, i, :], kind="quadratic")
+            p2.aSR[j, i, :] = f(p2.myt)
+            if p2.slip1[j, i] < 0:
+                p2.aSR[j, i, :] = 0
+                continue
```

Comment:
Why do we do this? Is not completly clear to me. Maybe add a comment.

Reply:
After the cubic interpolation, slip1 may get negative which does not make sense.